### PR TITLE
James c

### PR DESCRIPTION
**Pull request by jamesjamcow**
Changed the css for 
Signup.html
Login.html 

TO-DO 
fix the q and a section 
add styling to the home page

### CODE REVIEW

File: core/forms.py [hunk 1]
```diff
@@ -105,6 +117,80 @@ def __init__(self, *args, **kwargs):
...
 
+class HackerSettingForm(forms.ModelForm):
+    country = forms.CharField(max_length=255)
+    city = forms.CharField(max_length=255)
+    school = forms.CharField(max_length=255, required=False)
+    education = forms.ChoiceField(
+        choices=HACKATHON_EDUCATION_CHOICES, widget=forms.Select
+    )
+    username = forms.CharField(max_length=255)
+    email = forms.CharField(max_length=255)
```
Comment (JasonLovesDoggo) [inline core/forms.py]: Not as simple as just email. that needs to verified separately 
https://docs.allauth.org/en/dev/account/configuration.html

File: core/forms.py [hunk 2]
```diff
@@ -105,6 +117,80 @@ def __init__(self, *args, **kwargs):
             visible.field.widget.attrs["class"] = "form-button-style py-2"
 
 
+class HackerSettingForm(forms.ModelForm):
+    country = forms.CharField(max_length=255)
+    city = forms.CharField(max_length=255)
+    school = forms.CharField(max_length=255, required=False)
```
Comment (JasonLovesDoggo) [inline core/forms.py]: This should be a choicefield with the options being a school model

File: core/forms.py [hunk 3]
```diff
@@ -105,6 +117,80 @@ def __init__(self, *args, **kwargs):
...
+    birthday = forms.DateField(widget=forms.TextInput(attrs={"type": "date"}))
+    personal_website = forms.CharField(max_length=255, required=False)
+
+    class Meta:
+        model = Hacker
+        exclude = [
+            "objects",
+            "saved",
+            "saved_categories",
+            "notification_policy",
```
Comment (JasonLovesDoggo) [inline core/forms.py:154]: They should be able to config notif policy but potentially out of scope for this commit

File: core/models.py [hunk 4]
```diff
@@ -129,6 +129,10 @@ class Hacker(AbstractUser):
         null=True,
         help_text="Name of your school or university",
     )
+    # James Added the two
```
Comment (JasonLovesDoggo) [inline core/models.py]: delete that line

File: core/models.py [hunk 5]
```diff
@@ -261,15 +270,18 @@ class ReviewStatus(models.TextChoices):
 
 class Hackathon(MetaDataMixin):
     objects = HackthonsManager()
+
+    # This ID is to make it easier to identify hackathons when scraping in order to avoid duplicates
+    id = models.CharField(max_length=32, primary_key=True, editable=False)
```
Comment (JasonLovesDoggo) [inline core/models.py]: add unique=True

File: core/models.py [hunk 6]
```diff
@@ -369,6 +417,14 @@ class Meta:
     def __str__(self):
         return self.name
 
+    def get_id(cls, name, date):
+        return f"{name.lower().replace(' ', '_')}-{date.year}"
```
Comment (JasonLovesDoggo) [inline core/models.py]: If you're going to do it this way, set id to a https://docs.djangoproject.com/en/5.1/ref/models/fields/#generatedfield

File: core/views.py [hunk 7]
```diff
@@ -192,6 +195,6 @@ def is_admin(user):
 
 
 @user_passes_test(is_admin)
```
Comment (JasonLovesDoggo) [inline core/views.py]: even though you have the check, i would still add a ratelimter

File: core/models.py [hunk 8]
```diff
@@ -313,10 +325,14 @@ class Hackathon(MetaDataMixin):
 
     reimbursements = models.CharField(max_length=255, blank=True, null=True)
 
-    location = models.ForeignKey(
+    location = models.CharField(max_length=255, blank=True, null=True)
```
Comment (JasonLovesDoggo) [inline core/models.py]: validation? What's the point of a duplicated location field

File: core/models.py [hunk 9]
```diff
@@ -345,16 +361,48 @@ class Hackathon(MetaDataMixin):
...
+        max_length=1,
+        choices=HYBRID_CHOICES,
+        default="I",
+        help_text="Location of the hackathon, I for in-person, V for virtual, H for hybrid",
+    )
+
+    # Boolean fields for whether the hackathon is specific to certain groups
+    is_web3 = models.BooleanField(default=False)  # Web 3 hackathons
+    is_diversity = models.BooleanField(
+        default=False
```
Comment (JasonLovesDoggo) [inline core/models.py:380]: Ideally all of these actually contain the restrictions

File: core/models.py [hunk 10]
```diff
@@ -345,16 +361,48 @@ class Hackathon(MetaDataMixin):
...
+    custom_info = models.JSONField(
+        default=dict, null=True, blank=True
+    )  # Anything else that we might want to add in a structured format
+
     class Meta:
         ordering = ["start_date"]
 
         constraints = [
             models.CheckConstraint(  # ensure start date is before end date
-                check=models.Q(start_date__lt=models.F("end_date")),
```
Comment (JasonLovesDoggo) [inline core/models.py:343]: ehh... semantics but you generally don't want a start and end date (down to min iirc) to be at the same time.

**Review by JasonLovesDoggo — CHANGES_REQUESTED** [review]

File: core/models.py (hunk 10 above)
Comment (jamesjamcow) [reply core/models.py:343]: I mean the date is feb 12  2024 to feb 12 2024, what should I do ?

Comment (JasonLovesDoggo) [reply core/models.py:343]: Notice how it's a DateTimeField and not just a DateField.

**Comment by Bobliuuu** [comment]
this has to be a world record for longest commit on Github

**Pull request merged**